Source/DafnyCore/Resolver.cs

```diff
@@ -9941,6 +9944,51 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
       return true;
     }
 
+    bool CheckIfEqualityIsDefinitelyNotSupported(IndDatatypeDecl dt, Graph<IndDatatypeDecl/*!*/>/*!*/ dependencies) {
+
+      var scc = dependencies.GetSCC(dt);
+
+      foreach (var ctor in dt.Ctors) {
+
+        if (ctor.IsGhost) {
+          return true;
+        }
+
+        foreach (var arg in ctor.Formals) {
+          var anotherIndDt = arg.Type.AsIndDatatype;
+          if (arg.IsGhost ||
+              (anotherIndDt != null && anotherIndDt.EqualitySupport == IndDatatypeDecl.ES.Never) ||
+              arg.Type.IsCoDatatype ||
+              arg.Type.IsArrowType ||
+              (arg.Type.IsOpaqueType && !arg.Type.SupportsEquality)) {
```

Comment:
Also, I think using `arg.Type` here isn't safe; you want to expand the type first to make sure there's no subset type hiding the opaque type.